lib/resty/upstream/healthcheck.lua

```diff
@@ -296,39 +298,43 @@ local function check_peer_range(ctx, from, to, peers, is_backup)
 end
 
 local function check_peers(ctx, peers, is_backup)
-    local n = #peers
+    local n = 0
+    for _ in pairs(peers) do n = n + 1 end
     if n == 0 then
         return
     end
 
     local concur = ctx.concurrency
     if concur <= 1 then
-        for i = 1, n do
-            check_peer(ctx, i - 1, peers[i], is_backup)
+        for _, peer in pairs(peers) do
+            check_peer(ctx, peer.id, peer, is_backup)
         end
     else
         local threads
         local nthr
 
         if n <= concur then
-            nthr = n - 1
+            nthr = n
             threads = new_tab(nthr, 0)
-            for i = 1, nthr do
+
+            for k, peer in pairs(peers) do
+                if next(peers, _) == nil then
```

Comment:
checking the last peer with the current thread. 
because `peers` cant be indexed `peers[n]` returns nil